ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureBlobDataAccessor.java

```diff
@@ -418,43 +417,41 @@ public CompletableFuture<UpdateResponse> updateBlobMetadataAsync(BlobId blobId,
     for (List<CloudBlobMetadata> batchOfBlobs : partitionedLists) {
-      CompletableFuture<Void> deleteBatchOperationFuture = new CompletableFuture<>();
-      FutureUtils.orTimeout(storageClient.deleteBatch(batchOfBlobs), batchTimeout)
-          .whenComplete((responseList, throwable) -> {
-            if (throwable != null) {
-              deleteBatchOperationFuture.completeExceptionally(Utils.extractFutureExceptionCause(throwable));
-            } else {
-              for (int j = 0; j < responseList.size(); j++) {
-                Response<Void> response = responseList.get(j);
-                CloudBlobMetadata blobMetadata = batchOfBlobs.get(j);
-                // Note: Response.getStatusCode() throws exception on any error.
-                try {
-                  response.getStatusCode();
-                } catch (BlobStorageException bse) {
-                  int statusCode = bse.getStatusCode();
-                  // Don't worry if blob is already gone
-                  if (statusCode != HttpURLConnection.HTTP_NOT_FOUND && statusCode != HttpURLConnection.HTTP_GONE) {
-                    logger.error("Deleting blob {} got status {}", blobMetadata.getId(), statusCode);
-                    deleteBatchOperationFuture.completeExceptionally(bse);
-                  }
+      CompletableFuture<Void> currentBatchDeleteFuture =
+          FutureUtils.orTimeout(storageClient.deleteBatch(batchOfBlobs), batchTimeout).thenAccept(responses -> {
+            for (int j = 0; j < responses.size(); j++) {
+              Response<Void> response = responses.get(j);
+              CloudBlobMetadata blobMetadata = batchOfBlobs.get(j);
+              // Note: Response.getStatusCode() throws exception on any error.
+              try {
+                response.getStatusCode();
+              } catch (BlobStorageException bse) {
+                int statusCode = bse.getStatusCode();
+                // Don't worry if blob is already gone
+                if (statusCode != HttpURLConnection.HTTP_NOT_FOUND && statusCode != HttpURLConnection.HTTP_GONE) {
+                  logger.error("Deleting blob {} got status {}", blobMetadata.getId(), statusCode);
+                  throw new CompletionException(bse);
                 }
-                deletedBlobs.add(blobMetadata);
               }
-              // Complete the results for current iteration.
-              deleteBatchOperationFuture.complete(null);
+              deletedBlobs.add(blobMetadata);
             }
           });
-      deleteBatchOperationFutureList.add(deleteBatchOperationFuture);
+      if (resultFuture == null) {
+        resultFuture = currentBatchDeleteFuture;
+      } else {
+        // Issue batch deletes serially one after another since issuing all in parallel can be expensive resulting in 429s.
+        resultFuture = resultFuture.thenCompose(unused -> currentBatchDeleteFuture);
```

Comment:
This seems to defeat the purpose of passing a function that would return a CompletionStage.

The idea is to not send the delete request unless the first delete batch request has returned, right? But creating a CompletableFuture on each delete batch in this for loop would trigger delete request be sent out right away. We should do something like this

```
CompletableFuture<Void> resultFuture = CompletableFuture.completedFuture(null);
for (List<CloudBlobMetadata> batchOfBlobs : partitionedLists) {
    resultFuture = resultFuture.thenCompose(unused -> FutureUtils.orTimeout(....));
}
```

In this way, each delete batch would be encapsulated in the Function that is passed to thenCompose, so only when the previous delete batch response comes back , we would start composing the next completable future.